feat(connector):Add ironrdp-vmconnector crate, integrated into existing ironrdp-client and ironrdp async

irvingoujAtDevolution opened the pull request:
Hello @CBenoit , I added dynamically dispatched trait for hyperv connector and normal connectors. The main drive is to avoid copy and paste in previous version. I did it this is way is also to reuse existing code as much as possible, instead of copy each and every method and add a _vmconnect suffix. This might not be the best solution, let me know your thoughts.

CBenoit commented on crates/ironrdp-connector/Cargo.toml:
```diff
@@ -22,11 +22,15 @@ arbitrary = ["dep:arbitrary"]
 ironrdp-error = { path = "../ironrdp-error", version = "0.1" } # public
-ironrdp-pdu = { path = "../ironrdp-pdu", version = "0.5", features = ["std"] } # public
+ironrdp-pdu = { path = "../ironrdp-pdu", version = "0.5", features = [
+    "std",
+] } # public
 arbitrary = { version = "1", features = ["derive"], optional = true } # public
 sspi = "0.15" # public
 url = "2.5" # public
-rand_core = { version = "0.6", features = ["std"] } # TODO: dependency injection?
+rand_core = { version = "0.6", features = [
+    "std",
+] } # TODO: dependency injection?
```
nitpick: Revert this change since it’s unrelated with the current PR, and this is also not particularly a common style among the Rust community

CBenoit commented on crates/ironrdp-vmconnect/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+repository.workspace = true
+authors.workspace = true
+keywords.workspace = true
+categories.workspace = true
```
issue: Disable publish for now, otherwise release-plz will try to publish immediately.

```suggestion
categories.workspace = true
publish = false
```

CBenoit reviewed (commented):
question: What about the extra option `EnhancedMode`? I think Hyper-V understands the following format now: `<GUID>;EnhancedMode=1`

CBenoit commented on crates/ironrdp-client/src/rdp.rs:
```diff
@@ -287,79 +317,104 @@ where
     let mut buf = WriteBuf::new();
 
-    info!("Begin connection procedure");
+    info!(?pcb, "Begin connection procedure");
```
issue: Let’s not log the PCB Debug representation as an INFO-level log. You can log it in DEBUG in you want.

irvingoujAtDevolution replied:
updated

irvingoujAtDevolution commented:
We were having CredSSP issues on enhanced mode, I intend to have that support in separate pull request.

irvingoujAtDevolution commented on crates/ironrdp-client/src/config.rs:
```diff
@@ -238,6 +239,36 @@ struct Args {
+
+    /// The ID for the HyperV VM server to connect to
+    #[clap(long, value_parser)]
+    vmconnect: Option<String>,
```
updated

irvingoujAtDevolution commented on crates/ironrdp-client/src/config.rs:
```diff
@@ -351,12 +382,22 @@ impl Config {
 
+        let pcb = match (args.vmconnect, args.pcb) {
+            (Some(_), Some(_)) => {
+                anyhow::bail!("Cannot use both --vmconnect and --pcb options. Choose one of them.");
```
updated

irvingoujAtDevolution commented on crates/ironrdp-connector/src/lib.rs:
```diff
@@ -382,3 +384,49 @@ where
+pub trait ConnectorCore: Sequence + SecurityConnector + CredsspSequenceFactory + Any {
+    fn as_sequence(&mut self) -> &mut dyn Sequence;
+    fn as_security_connector(&mut self) -> &mut dyn SecurityConnector;
+    fn into_any(self: Box<Self>) -> Box<dyn Any>;
```
updated

CBenoit commented:
@irvingoujAtDevolution Can you double check with `cargo xtask ci`?

CBenoit commented:
@irvingoujAtDevolution Can you resolve the conflicts with master?

CBenoit commented on Cargo.toml:
```diff
@@ -210,3 +210,4 @@ opt-level = 3
 # FIXME: We need to catch up with Diplomat upstream again, but this is a significant amount of work.
 # In the meantime, we use this forked version which fixes an undefined behavior in the code expanded by the bridge macro.
 diplomat = { git = "https://github.com/CBenoit/diplomat", rev = "6dc806e80162b6b39509a04a2835744236cd2396" }
```
@irvingoujAtDevolution Most likely, you can remove the git patch here.

irvingoujAtDevolution committed:
feat: add vmconnect support with VmClientConnector

richerlariviere reviewed (approved):
I see no change to the .github so DevOps review is irrelevant for this PR. I approve but in this case I dismiss this request review from us.

CBenoit reviewed (changes requested):
I’m sorry; just making sure this does not get merged before we are done with the current publishing batch.

irvingoujAtDevolution committed:
refactor(vmconnect): simplify connector composition

Replace the type-erased `Box<dyn ConnectorCore>` + `Any` downcast round-trip
with explicit static branches per transport, and drop the parallel CredSSP
machinery the previous approach required.

- remove the `ConnectorCore`, `CredsspSequenceFactory`, and
  `CredsspSequenceTrait` traits, and the `Any`-downcast recovery of the
  concrete connector after CredSSP
- vmconnect now reuses the standard `CredsspSequence` (NTLM via
  `protocol = HYBRID`, no Kerberos config) instead of a parallel
  `VmCredsspSequence`
- `SecurityConnector: Sequence` is the only remaining trait; rename
  `selected_protocol` -> `credssp_protocol` so no impl has to fake a value
- move `PreconnectionBlobPayload` into ironrdp-connector (single definition,
  re-exported from ironrdp-client)
- ironrdp-async no longer depends on ironrdp-vmconnect: `run_until_handover`
  moves into ironrdp-vmconnect, `send_pcb`/`PcbSent` stay in ironrdp-async
- `VmClientConnector::take_over` takes `PcbSent` and returns `Result`
  instead of panicking; drop the no-op state transitions driven by fake
  `step(&[])` calls

irvingoujAtDevolution committed:
fix(vmconnect): harden connection sequence and address review findings

Correctness:
- Reject instead of panic when a standard-path RDCleanPath Response omits the X.224
  connection confirm: it previously skipped the connector step and then tripped
  skip_connect_begin's assertion, a peer-reachable panic (native and web paths).
- Require the host to select exactly PROTOCOL_HYBRID for a vmconnect (Direct Approach)
  session; SSL or HYBRID_EX are now rejected rather than echoed into Client Core Data
  as a protocol we never performed.
- Map vmconnect negotiation failures to ConnectorErrorKind::Negotiation, matching the
  standard connector so downstream error handling recognizes them.
- Preserve the destination port through the RDS Gateway transport; ironrdp-mstsgu no
  longer hard-codes 3389, so a non-3389 target (including a Hyper-V host) is reachable.
- Send a general preconnection blob ahead of the handshake on the direct and gateway
  transports too, not only through RDCleanPath.

API and hygiene:
- Drop the dead network-client-less CredSSP branch (perform_credssp takes &mut N),
  the redundant VmClientConnector fields, and the unused PreconnectionBlobPayload::vmconnect().
- Narrow the ironrdp-vmconnect public surface to VmClientConnector + run_until_handover.
- Rename CredSSPFinished -> CredsspFinished (RFC 430); mark PreconnectionBlobPayload
  #[non_exhaustive]; restore the # Panics docs on the SecurityConnector trait methods.

BREAKING CHANGE: ironrdp-connector moves ClientConnector's security-phase methods
(should_perform_security_upgrade, mark_security_upgrade_as_done, should_perform_credssp,
mark_credssp_as_done) onto the new SecurityConnector trait, so callers must import it;
adds PreconnectionBlobPayload (non_exhaustive). ironrdp-async splits connect_finalize into
perform_credssp + connect_finalize with a new CredsspFinished token. ironrdp-mstsgu's
GwConnectTarget gains a required `port` field. ironrdp-rdcleanpath's RDCleanPathPdu::new_request
takes an optional X.224 PDU and RDCleanPath::Response carries an optional X.224 connection response.

irvingoujAtDevolution commented:
Close in favor of new design
